src/rdf4cpp/rdf/parser/SerdIteratorImpl.ipp

```diff
@@ -0,0 +1,336 @@
+#include <rdf4cpp/rdf/parser/IStreamQuadIterator.hpp>
+
+#include <cassert>
+#include <cstddef>
+#include <deque>
+#include <istream>
+#include <utility>
+
+#if __has_include(<serd/serd.h>)
+#include <serd/serd.h>
+#else
+#include <serd-0/serd/serd.h>
+#endif
+
+
+#include <rdf4cpp/rdf/storage/util/robin-hood-hashing/robin_hood_hash.hpp>
+#include <rdf4cpp/rdf/Quad.hpp>
+
+namespace rdf4cpp::rdf::parser {
+
+struct IStreamAdaptor {
+    std::reference_wrapper<std::istream> istream;
+    uint32_t line_off = 0;
+    uint32_t col_off = 0;
+
+    template<bool track_position>
+    static size_t read(void *buf, size_t elem_size, size_t count, void *voided_self) {
+        assert(elem_size == 1);
+        assert(count == 1);
+
+        auto *self = reinterpret_cast<IStreamAdaptor *>(voided_self);
+        self->istream.get().read(reinterpret_cast<char *>(buf), static_cast<std::streamsize>(count));
+
+        if constexpr (track_position) {
+            if (static_cast<char const *>(buf)[0] == '\n') {
+                self->line_off += 1;
+                self->col_off = 0;
+            }
+        }
+
+        return self->istream.get().gcount();
+    }
+
+    static int is_ok(void *voided_self) {
+        auto *self = reinterpret_cast<IStreamAdaptor *>(voided_self);
+        return self->istream.get() ? 0 : 1;
+    }
+};
+
+struct IStreamQuadIterator::Impl {
```

Comment:
rename to e.g. `IStreamQuadIteratorSerdBackend`. Rename file accordingly.  Move it to namespace/folder `backend` (I am open for better name suggestions.